src/controllers/auth-controller.spec.ts

```diff
@@ -430,4 +447,192 @@ describe('Auth', () => {
+    let firstUser: User;
+    let firstUserResetToken: string;
+
+    const secondUserEmail = 'reset-password-expired@mail.com';
+    const secondUserPassword = '45678';
+    let secondUser: User;
+    let expiredResetToken: string;
+
+    beforeAll(async () => {
+      [firstUser, secondUser] = await Promise.all(
+        [
+          { email: firstUserEmail, password: firstUserPassword },
+          { email: secondUserEmail, password: secondUserPassword },
+        ].map(creds => User.create(creds)),
+      );
+      await Promise.all([firstUser, secondUser].map(user => userRepository.save(user)));
+
+      firstUserResetToken = signResetToken({ email: firstUserEmail, userId: firstUser.id });
+      expiredResetToken = signResetToken({ email: secondUserEmail, userId: secondUser.id }, -1);
+
+      await Promise.all(
+        [{ user: firstUser, token: firstUserResetToken }, { user: secondUser, token: expiredResetToken }].map(
+          ({ user, token }) => resetTokenRepository.save(new ResetToken(token, user.id)),
+        ),
+      );
+    });
+
+    beforeEach(() => {
+      spyOnSendResetCompleteMail.mockClear();
+    });
+
+    it('not be able update password without new password provided', async () => {
+      const res = await request(app)
+        .post(urls.reset)
+        .set('Accept', 'application/json')
+        .send({ token: firstUserResetToken });
+      const token = await resetTokenRepository.findOne({ token: firstUserResetToken });
+      expect(res.status).toEqual(400);
+      expect(token).toBeDefined();
+      expect(token!.token).toEqual(firstUserResetToken);
+      expect(JSON.parse(res.text).error).toEqual('Reset token and passwords are required');
+      expect(spyOnSendResetCompleteMail).not.toHaveBeenCalled();
+    });
+
+    it('not be able update password without reset token provided', async () => {
+      const res = await request(app)
+        .post(urls.reset)
+        .set('Accept', 'application/json')
+        .send({ password: firstUserPassword });
+      const token = await resetTokenRepository.findOne({ token: firstUserResetToken });
+      expect(res.status).toEqual(400);
+      expect(token).toBeDefined();
+      expect(token!.token).toEqual(firstUserResetToken);
+      expect(JSON.parse(res.text).error).toEqual('Reset token and passwords are required');
+      expect(spyOnSendResetCompleteMail).not.toHaveBeenCalled();
+    });
+
+    it('be able to update password with passed token and new password; and by this disable used token', async () => {
+      const response = await request(app)
+        .post(urls.reset)
+        .set('Accept', 'application/json')
+        .send({ token: firstUserResetToken, password: firstUserNewPassword });
+
+      const sameToken = await resetTokenRepository.findOne({ userId: firstUser.id });
+      const sameUser = await userRepository.findOne({ id: firstUser.id });
+
+      if (!sameUser) {
+        throw new Error('user should exist');
+      }
```

Comment:
This is a controversial check, it roughly means that it is possible that after a user requested a reset and then came to enter a new password, someone or something managed to remove it from the database.

The fact of the user's existence is confirmed when requesting to reset the password.

I do not deny the possibility of such a case, but it seems to me too incredible. However, this check does not verify anything...

Reply:
That was done just to get rid of `undefined` in the type of `dbUser`. So, I guess I can simply remove it and use `!` below.